File: sorting/bubble_sort/bubble_sort.js
function bubble_sort(elements) {
    for (let i = 0; i < elements.length; i++) {
        for (let j = 0; j < elements.length - 1; j++) {
            if (elements[j] > elements[j + 1]) {
                [elements[j], elements[j + 1]] = [elements[j + 1], elements[j]];
            }
        }
    }
    return elements;
}

function bubble_sort_optimized(elements) {
    var i, j;
    var swapped;
    const length = elements.length;

    for (i = 0; i < length - 1; i++) {
        for (j = 0; j < length - 1 - i; j++) {
            if (elements[j] > elements[j + 1]) {
                [elements[j], elements[j + 1]] = [elements[j + 1], elements[j]];
                swapped = true;
            }
        }
        if (swapped == false) break;
    }
    return elements;
}

const input = [3, 7, 2, 9, 0, 6]

console.log(`bubble_sort: ${bubble_sort(input.slice())}`);
console.log(`bubble_sort_optimized: ${bubble_sort(input.slice())}`);

File: sorting/bubble_sort/bubble_sort.py
def bubble_sort(elements: list[int]) -> list[int]:

    for _ in range(len(elements)):
        for j in range(len(elements) - 1):
            if elements[j] > elements[j+1]:
                elements[j], elements[j+1] = elements[j+1], elements[j]

    return elements


def bubble_sort_optimized(elements: list[int]) -> list[int]:
    length: int = len(elements)
    swapped: bool = False

    for i in range(length - 1):
        swapped = False
        for j in range(length - 1 - i):
            if elements[j] > elements[j+1]:
                elements[j], elements[j+1] = elements[j+1], elements[j]
                swapped = True

        if not swapped:
            break

    return elements


if __name__ == '__main__':
    input = [8, 2, 2, 9, 1]
    print(input)
    print(f"{bubble_sort(input[:]) = }")
    print(f"{bubble_sort_optimized(input[:]) = }")
